### ISSUE TITLE
Write evdns wrappers

### ISSUE THREAD

**Issue by bassosimone** [bassosimone/mkok-libevent-ng#1]
Reposting here bassosimone/mkok-libevent#1:

---

Rather than writing what should be done, here I'd go one step further and will directly write the documentation of what should be implemented. To complete the issue, also add documentation for the implemented functionality, starting from what is outlined below.

---
# Wrappers for evdns

Wrappers for evdns live in the `include/mkok/libevent.hpp` file and are static functions inside the `EvdnsBase` class. The following functions are implemented:
## EvdnsBase::create()

This function creates a new `evdns_base` shared pointer. It has the following signatures:

``` C++
Var<evdns_base> EvdnsBase::create(Var<event_base> base,
        bool initialize_nameservers = true,  bool fail_requests = true);
```

Its implementation shall perform the following actions:
1. call `evdns_base_new()` and throw `EvdnsBaseNewException` in case of failure
2. make sure that `base` is referenced as long as the to-be-returned `Var<evdns_base>` is referenced by some code
3. make sure that `evdns_base_free()` is called with the proper `fail_requests` argument when the last reference to the returned `Var<evdns_base>` is gone.

To write the implementation, you could see how `EventBase::create()` is implemented. Also, it would help to peek at [relight implementation of resolver](https://github.com/bassosimone/relight/blob/master/dns-resolver.hpp).
## EvdnsBase::resolve_ipv4()

This functions performs `A` resolution of a domain name. It has the following signature:

``` C++
typedef std::function<void(int result, char type, int count, int ttl,
        std::vector<std::string> addresses) ResolveCallback;

void EvdnsBase::resolve_ipv4(Var<evdns_base> base, std::string name,
        ResolveCallback callback, int flags = DNS_QUERY_NO_SEACH);
```

Its implementation shall perform the following actions:
1. make sure that `callback` and its closure are alive for long enough making an heap copy of it, using a technique similar to the one used by `EventBase::once()`, and of course also making sure that there are no memory leaks on failure
2. call `evdns_base_resolve_ipv4()` with the proper arguments and raise `EvdnsBaseResolveIpv4Exception` if this function returns an error
3. in the callback (which must be implemented as an `extern "C"` function to ensure maximum compatibility with C code) it must process the result to store it inside a `std::vector<std::string>` using code similar to the one used by measurement-kit, and then call the callback and make sure there are no resource leaks

To write the implementation you could also check how [a similar task is performed by relight](https://github.com/bassosimone/relight/blob/master/dns.cpp).
## EvdnsBase::resolve_ipv6()

This is equal to `EvdnsBase::resolve_ipv4()` except that it performs `AAAA` queries. To implement, I suggest to check how this is done by measurement-kit.
## EvdnsBase::resolve_reverse()

This is equal to `EvdnsBase::resolve_ipv4()` except that it performs a `PTR` query. To implement, I suggest to check how this is done by measurement-kit.
## EvdnsBase::resolve_reverse_ipv6()

This is equal to `EvdnsBase::resolve_ipv4()` except that it performs a `PTR` query. To implement, I suggest to check how this is done by measurement-kit.
## EvdnsBase::clear_nameservers_and_suspend()

This function clears all the nameservers configured inside an `evdns_base` and suspends any resolve operation until `base_resume()` is called. It has the following signature:

``` C++
void EvdnsBase::clear_nameservers_and_suspend(
        Var<evdns_base> base);
```

Its implementation shall perform the following actions:
1. call `evdns_base_clear_nameservers_and_suspend()` and throw `EvdnsBaseClearNameserversAndSuspendException` is that fails

An implementation of this already exists in measurement-kit [src/common/poller.cpp](https://github.com/measurement-kit/measurement-kit/blob/master/src/common/poller.cpp#L44).
## EvdnsBase::count_nameservers

This function counts all the nameservers inside an `evdns_base`. It has the following signature:

``` C++
unsigned EvdnsBase::count_nameservers(Var<evdns_base> base);
```

Its implementation shall perform the following actions:
1. call `evdns_base_count_nameservers()` and raise `EvdnsBaseCountNameserversException` if that fails
2. return the number of nameservers

Again, this is already implemented in measurement-kit [src/common/poller.cpp](https://github.com/measurement-kit/measurement-kit/blob/master/src/common/poller.cpp#L51).
## EvdnsBase::add_nameserver

This function adds a nameserver (expressed as an address optionally followed by a colon and by a port number) to the nameservers inside an `evdns_base`. It has the following signature:

``` C++
void EvdnsBase::add_nameserver(Var<evdns_base> base,
        std::string nameserver);
```

Its implementation shall perform the following actions:
1. call `EvdnsBase::nameserver_ip_add()` and throw `EvdnsBaseNameserverIpAddException` if that fails

An implementation of this is available in measurement-kit [src/dns/resolver.cpp](https://github.com/measurement-kit/measurement-kit/blob/master/src/dns/resolver.cpp#L53).
## EvdnsBase::resume()

This function resume operations after `clear_nameservers_and_suspend()` was called. It has the following signature:

``` C++
void EvdnsBase::resume(Var<evdns_base> base);
```

Its implementation shall perform the following actions:
1. call `evdns_base_resume()` and throw `EvdnsBaseResumeException` is that fails

An implementation of this already exists in measurement-kit [src/common/poller.cpp](https://github.com/measurement-kit/measurement-kit/blob/master/src/common/poller.cpp#L44).
## EvdnsBase::set_option_attempts()

This function configures the number of times that a request is retried before deciding that no response was received. It has the following signature:

``` C++
void EvdnsBase::set_option_attempts(Var<evdns_base> base, unsigned count);
```

Its implementation shall perform the following actions:
1. call `evdns_base_set_option()`, passing it the proper option name and value, and throw `EvdnsBaseSetOptionException` is that fails

An implementation of this already exists in measurement-kit [src/dns/resolver.cpp](https://github.com/measurement-kit/measurement-kit/blob/master/src/dns/resolver.cpp#L77).
## EvdnsBase::set_option_timeout()

This function configures the timeout after which a DNS request is declared failed. It has the following signature:

``` C++
void EvdnsBase::set_option_timeout(Var<evdns_base> base, double timeo);
```

Its implementation shall perform the following actions:
1. call `evdns_base_set_option()`, passing it the proper option name and value, and throw `EvdnsBaseSetOptionException` is that fails

An implementation of this already exists in measurement-kit [src/dns/resolver.cpp](https://github.com/measurement-kit/measurement-kit/blob/master/src/dns/resolver.cpp#L83).
## EvdnsBase::set_option_randomize_case()

This function configures whether we should send DNS requests with randomized case. It has the following signature:

``` C++
void EvdnsBase::set_option_randomize_case(Var<evdns_base> base,
        bool yesno);
```

Its implementation shall perform the following actions:
1. call `evdns_base_set_option()`, passing it the proper option name and value, and throw `EvdnsBaseSetOptionException` is that fails

An implementation of this already exists in measurement-kit [src/dns/resolver.cpp](https://github.com/measurement-kit/measurement-kit/blob/master/src/dns/resolver.cpp#L95).

**Comment by bassosimone** [comment]
(Optionally, evdns wrappers could be part of another library)

**Comment by bassosimone** [comment]
Ping @AntonioLangiu

**Comment by AntonioLangiu** [comment]
How can I see if `evdns_base_count_nameservers()` fails? In the [official documentation](http://www.wangafu.net/~nickm/libevent-2.1/doxygen/html/dns_8h.html#a6811354521b4fe8df4c724d6cb9fb102) is not reported!

I think a possible error can be -1, but I'm not sure.

**Comment by bassosimone** [comment]
I don't think it can fail, but to be safe I'd throw if the result is negative

**Comment by AntonioLangiu** [comment]
:+1:
